### PR TITLE
Limits Header extension: add support for metric hierarchies with more than 2 levels

### PR DESCRIPTION
**Pull request by davidor**
This is part of #114 . It adapts the limits header extension to work with metric hierarchies of more than 2 levels.

### CODE REVIEW

File: lib/3scale/backend/metric.rb [hunk 1]
```diff
@@ -162,6 +162,17 @@ def descendants(service_id, metric_name)
           end
         end
 
+        # Returns the "ascendants" of a metric, that is, its parents,
```
Comment (unleashed) [inline lib/3scale/backend/metric.rb]: can a metric have multiple parents?

File: lib/3scale/backend/metric.rb [hunk 2]
```diff
@@ -162,6 +162,17 @@ def descendants(service_id, metric_name)
...
+        # grandparents, etc. The "ascendants" of a metric are its parents plus
+        # the ascendants of each of them.
+        def ascendants(service_id, metric_name)
+          parents_of_metric = parents(service_id, [metric_name])
+
+          parents_of_metric.reduce(parents_of_metric) do |acc, parent|
+            acc + ascendants(service_id, parent)
+          end
```
Comment (unleashed) [inline lib/3scale/backend/metric.rb]: Have you considered using `parents_of_metric.each` and pushing to it as you go for a more readable and performant version avoiding recursion?

File: test/unit/metric_test.rb [hunk 3]
```diff
@@ -261,4 +261,22 @@ def test_descendants
...
+  def test_ascendants
+    service_id = next_id
+    Service.save!(provider_key: 'a_provider_key', id: service_id)
+    levels = rand(3..10)
+    metrics = gen_hierarchy_one_metric_per_level(service_id, levels)
+
+    assert_true metrics.each_with_index.all? do |metric, idx|
+      ascendants = Metric.ascendants(service_id, metric.name)
+
+      if idx == 0
```
Comment (davidor) [inline test/unit/metric_test.rb]: This can be simplified using `take(idx-1)`.

Comment (unleashed) [reply test/unit/metric_test.rb]: yes, are you changing it?

File: lib/3scale/backend/metric.rb (hunk 1 above)
Comment (davidor) [reply lib/3scale/backend/metric.rb]: No.

Comment (unleashed) [reply lib/3scale/backend/metric.rb]: yep, this was meant to correct the parents to `parent, grandparent, ...`

File: lib/3scale/backend/metric.rb (hunk 2 above)
Comment (davidor) [reply lib/3scale/backend/metric.rb]: Not sure I follow. Can you expand a bit?

File: lib/3scale/backend/metric.rb (hunk 1 above)
Comment (davidor) [reply lib/3scale/backend/metric.rb]: Right. The description is not accurate :+1:

File: lib/3scale/backend/metric.rb (hunk 2 above)
Comment (davidor) [reply lib/3scale/backend/metric.rb]: :trollface:

Comment (unleashed) [reply lib/3scale/backend/metric.rb]: This is the exact same code as `descendants` - have you checked out how to DRY this up?

Comment (davidor) [reply lib/3scale/backend/metric.rb]: I haven't come up with anything that makes the code more readable.

File: test/unit/metric_test.rb (hunk 3 above)
Comment (davidor) [reply test/unit/metric_test.rb]: Done :+1:

File: lib/3scale/backend/metric.rb (hunk 1 above)
Comment (davidor) [reply lib/3scale/backend/metric.rb]: Fixed :+1:

File: lib/3scale/backend/metric.rb (hunk 2 above)
Comment (unleashed) [reply lib/3scale/backend/metric.rb]: This makes me feel uneasy, but oh well.

**Comment by davidor** [comment]
bors r=@unleashed

**Pull request merged**